beacon_node/eth1/src/deposit_cache.rs

```diff
@@ -114,33 +175,111 @@ impl DepositCache {
     pub fn is_empty(&self) -> bool {
-        self.logs.is_empty()
+        self.finalized_deposit_count != 0 && self.logs.is_empty()
     }
 
     /// Returns the block number for the most recent deposit in the cache.
-    pub fn latest_block_number(&self) -> Option<u64> {
-        self.logs.last().map(|log| log.block_number)
+    pub fn latest_block_number(&self) -> u64 {
+        self.logs
+            .last()
+            .map(|log| log.block_number)
+            .unwrap_or(self.finalized_block_height)
     }
 
-    /// Returns an iterator over all the logs in `self`.
+    /// Returns an iterator over all the logs in `self` that aren't finalized.
     pub fn iter(&self) -> impl Iterator<Item = &DepositLog> {
         self.logs.iter()
     }
 
-    /// Returns the i'th deposit log.
-    pub fn get(&self, i: usize) -> Option<&DepositLog> {
-        self.logs.get(i)
+    /// Returns the deposit log with INDEX i.
+    pub fn get_log(&self, i: usize) -> Option<&DepositLog> {
+        let finalized_deposit_count = self.finalized_deposit_count as usize;
+        if i < finalized_deposit_count {
+            None
+        } else {
+            self.logs.get(i - finalized_deposit_count)
+        }
+    }
+
+    /// Returns the deposit root with DEPOSIT COUNT (not index) i
+    pub fn get_root(&self, i: usize) -> Option<&Hash256> {
+        let finalized_deposit_count = self.finalized_deposit_count as usize;
+        if i < finalized_deposit_count {
+            None
+        } else {
+            self.deposit_roots.get(i - finalized_deposit_count)
+        }
+    }
+
+    /// Returns the finalized deposit count
+    pub fn finalized_deposit_count(&self) -> u64 {
+        self.finalized_deposit_count
+    }
+
+    /// Finalizes the cache up to `eth1_block.deposit_count`.
+    pub fn finalize(&mut self, eth1_block: Eth1Block) -> Result<(), Error> {
+        let deposits_to_finalize = eth1_block.deposit_count.ok_or_else(|| {
+            Error::Internal("Eth1Block did not contain deposit_count".to_string())
+        })?;
+
+        let currently_finalized = self.finalized_deposit_count;
+        if deposits_to_finalize > self.len() as u64 || deposits_to_finalize <= currently_finalized {
+            Err(Error::InvalidFinalizeIndex {
+                requested_count: deposits_to_finalize,
+                currently_finalized,
+                deposit_count: self.len() as u64,
+            })
+        } else {
+            let finalized_log = self
+                .get_log((deposits_to_finalize - 1) as usize)
+                .cloned()
+                .expect("log should exist");
+            let drop = (deposits_to_finalize - currently_finalized) as usize;
+            self.deposit_tree
+                .finalize(eth1_block.into())
+                .map_err(Error::DepositTree)?;
+            self.logs.drain(0..drop);
+            self.leaves.drain(0..drop);
+            self.deposit_roots.drain(0..drop);
```

Comment:
depending on how we handle the `deposit_roots` issue we may want to modify this `drain` to preserve the first element

Reply:
this is correct as it is :)

`get_root(index)` returns the same root before and after finalization :)

Reply:
awesome, thanks!